docs/arc42.md

```diff
@@ -0,0 +1,741 @@
+   - **Basic Message Protocol** will enable straightforward message exchanges between agents.
+
+3. **Future Development**:
+   - Support for **Acknowledgments (Acks)** and the **Present Proof Protocol** is planned, targeting comprehensive compliance with DIDComm standards.
+
+This modular and extensible implementation ensures the mediator evolves with the SSI ecosystem while maintaining high standards for security and performance.
+
+---
+
+## 5. Building Block View
+
+### 5.1 Overview
+
+The `didcomm-mediator-rs` project utilizes a plugin-based architecture, allowing modular and extensible implementation of DIDComm messaging protocols. The architecture is divided into two key plugin layers:
+
+1. **Web Plugin Layer**: 
+   - Enables plug-and-play of web endpoints.
+   - Currently used to implement features such as DID endpoints, out-of-band messages, and DIDComm messaging.
+
+2. **DIDComm Messaging Plugin Layer**:
+   - Facilitates incremental addition of DIDComm sub-protocols.
+   - Modular implementation ensures support for future protocols without major refactoring.
+
+### 5.2 Detailed View
+
+#### Top-Level Architecture
+
+At the core is the **Generic Server**, responsible for hosting the overall application and enabling the plugin system. The main components are:
+
+1. **Dispatcher**: Routes incoming requests to the appropriate web plugin.
+2. **Web Plugin Layer**: Hosts modular implementations for endpoint-specific logic (e.g., DID endpoint generation, out-of-band messages).
+3. **Common DIDComm Processing**: Handles general DIDComm request/response processing applicable across sub-protocols.
+
+#### DIDComm Messaging Plugin Layer
+
+Nested within the DIDComm Messaging endpoint is another plugin system for managing sub-protocols. This includes:
+
+1. **Forward Protocol**: Handles message routing across agents.
+2. **Pickup Protocol**: Manages message retrieval by offline agents.
+3. **Mediator Coordination Protocol**: Supports agent registration and mediation setup.
+4. **Plugin Utilities**: Shared utilities that facilitate sub-protocol implementations.
+
+### 5.3 Building Block Diagram
+
+The following diagram illustrates the layered architecture with the plugin-based system:
+
+```mermaid
+graph TD
+    GenericServer[Generic Server] --> Dispatcher[Dispatcher]
+    Dispatcher --> WebPluginLayer[Web Plugin Layer]
+    WebPluginLayer --> DIDEndpoint[DID Endpoint]
+    WebPluginLayer --> OOBMessaging[Out-of-Band Messages]
+    WebPluginLayer --> DIDCommMessaging[DIDComm Messaging Endpoint]
+    DIDCommMessaging --> CommonProcessing[Common DIDComm Processing]
+    DIDCommMessaging --> MessagingPlugins[Messaging Plugin Layer]
+    MessagingPlugins --> Forward[Forward Protocol]
+    MessagingPlugins --> Pickup[Pickup Protocol]
+    MessagingPlugins --> Coordination[Mediator Coordination Protocol]
+    MessagingPlugins --> Trustping[Trust Ping Protocol]
+```
```

Comment:
Please change the Out-of-Band messages to Out-of-Band Messaging

Reply:
done